cmd/util/cmd/epochs/cmd/recover.go

```diff
@@ -0,0 +1,186 @@
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/state/protocol/inmem"
+)
+
+// generateRecoverEpochTxArgsCmd represents a command to generate the data needed to submit an epoch recovery transaction the network is in EFM (epoch fallback mode).
+// EFM can be exited only by a special service event, EpochRecover, which initially originates from a manual service account transaction.
+// The full epoch data must be generated manually and submitted with this transaction in order for an
+// EpochRecover event to be emitted. This command retrieves the current protocol state identities, computes the cluster assignment using those
+// identities, generates the cluster QC's and retrieves the DKG key vector of the last successful epoch.
+var (
+	generateRecoverEpochTxArgsCmd = &cobra.Command{
+		Use:   "generate-efm-recovery-data",
+		Short: "Generates recover epoch transaction arguments",
+		Long:  "Generates transaction arguments for the epoch recovery transaction.",
+		Run:   generateRecoverEpochTxArgs,
+	}
+
+	flagAnAddress               string
+	flagAnPubkey                string
+	flagPartnerWeights          string
+	flagPartnerNodeInfoDir      string
+	flagInternalNodePrivInfoDir string
+	flagConfig                  string
+	flagCollectionClusters      int
+	flagStartView               uint64
+	flagStakingEndView          uint64
+	flagEndView                 uint64
+)
+
+func init() {
+	rootCmd.AddCommand(generateRecoverEpochTxArgsCmd)
+	addGenerateRecoverEpochTxArgsCmdFlags()
+}
+
+func addGenerateRecoverEpochTxArgsCmdFlags() {
+	generateRecoverEpochTxArgsCmd.Flags().StringVar(&flagBucketNetworkName, "bucket-network-name", "",
+		"when retrieving the root snapshot from a GCP bucket, the network name portion of the URL (eg. \"mainnet-13\")")
+	generateRecoverEpochTxArgsCmd.Flags().IntVar(&flagCollectionClusters, "collection-clusters", 0,
+		"number of collection clusters")
+	// required parameters for network configuration and generation of root node identities
+	generateRecoverEpochTxArgsCmd.Flags().StringVar(&flagConfig, "config", "",
+		"path to a JSON file containing multiple node configurations (fields Role, Address, Weight)")
+	generateRecoverEpochTxArgsCmd.Flags().StringVar(&flagInternalNodePrivInfoDir, "internal-priv-dir", "", "path to directory "+
+		"containing the output from the `keygen` command for internal nodes")
+	generateRecoverEpochTxArgsCmd.Flags().StringVar(&flagPartnerNodeInfoDir, "partner-dir", "", "path to directory "+
+		"containing one JSON file starting with node-info.pub.<NODE_ID>.json for every partner node (fields "+
+		" in the JSON file: Role, Address, NodeID, NetworkPubKey, StakingPubKey)")
+	generateRecoverEpochTxArgsCmd.Flags().StringVar(&flagPartnerWeights, "partner-weights", "", "path to a JSON file containing "+
+		"a map from partner node's NodeID to their stake")
+	generateRecoverEpochTxArgsCmd.Flags().Uint64Var(&flagStartView, "start-view", 0, "start view of the recovery epoch")
```

Comment:
## Start view sensitivity
The `startView` parameter is very sensitive, because in order for a `EpochRecover` service event to be valid, it must be exactly one view after the last `EpochExtension`. Otherwise we would have conflicting/missing definitions for who is the leader for a view. 

Suggestions:
- Retrieve the "extended final view" from the snapshot (`FinalView` from the latest extension of the current epoch)
- Validate that `--start-view` is equal to `extendedFinalView+1` 

Alternatively we can revisit the strictness of the requirement, but would need to carefully think through the safety implications. 

## Race condition
In addition, there is a race condition here to be aware of. The [design specifies](https://www.notion.so/flowfoundation/Spork-less-Epoch-Fallback-Recovery-Design-II-Epoch-Extensions-e923ca84b4d848c5a17de4b18fb5afd5?pvs=4#9f112daeda3244c98b20e1b861fa4042) that we will add a new extension periodically. If, in between retrieving the snapshot while executing this command and submitting the corresponding admin transaction, we add a new epoch extension, the `RecoverEpoch` event would be discarded as invalid. 

This isn't the end of the world because we can just do it again, but it might be worth having the tooling help avoid this situation, for example by reporting the number of view remaining before another extension would be added. To be honest I'm kind of conflicted: on the one hand it would be useful but on the other hand, this is already an infrequent, heavily involved process where I'd expect humans to be double-checking these things.

Reply:
This comment is related to the comment above it and the same commit applies https://github.com/onflow/flow-go/pull/5576/commits/884fdbf2dd8a1508f897819c55ea4e953287efa2 , as far as the race condition it's unlikely to happen and we would only be able to log a useful message if we were able to get the current view from the network somehow then we can could log how many views left until the start-view .